File: ModelJson/ModelJson/Renderer.m
//
//  Renderer.m
//  ModelJson
//
//  Created by Jacob Su on 4/26/21.
//

@import ModelIO;
@import MetalKit;
@import Metal;
#import <Foundation/Foundation.h>
#import "Renderer.h"
#import "ShaderType.h"
#import <common/common.h>

@implementation Renderer
{
    id<MTLDevice> _device;
    id<MTLRenderPipelineState> _pipelineState;
    id<MTLCommandQueue> _commandQueue;
    id<MTLDepthStencilState> _depthState;
    id<MetalMesh> _foxMesh;
    Uniforms _uniforms;
    vector_uint2 _viewportSize;
    id<Camera> _camera;
    SatelliteCameraController *_satelliteController;
}

- (nonnull instancetype)initWithMetalKitView:(nonnull MTKView*)mtkView
{
    self = [super init];
    
    if (self) {
        mtkView.delegate = self;
        mtkView.sampleCount = 4;
        id<MTLDevice> device = mtkView.device;
        _device = mtkView.device;
        _camera = [[SimpleCamera alloc] initWithPosition:(vector_float3) {5.0, 0.0, 8.0}
                                              withTarget:(vector_float3){0.0, 0.0, 0.0}
                                                      up:YES];
        _satelliteController = [[SatelliteCameraController alloc] initWithCamera:_camera];
        
        mtkView.depthStencilPixelFormat = MTLPixelFormatDepth32Float;
        mtkView.clearDepth = 1.0;
        MTLDepthStencilDescriptor *depthStencilDesc = [MTLDepthStencilDescriptor new];
        depthStencilDesc.depthCompareFunction = MTLCompareFunctionLess;
        depthStencilDesc.depthWriteEnabled = YES;
        _depthState = [device newDepthStencilStateWithDescriptor:depthStencilDesc];
        
        MTLVertexDescriptor *mtlVertexDescriptor = [[MTLVertexDescriptor alloc] init];
        // position
        mtlVertexDescriptor.attributes[ModelVertexAttributePosition].format = MTLVertexFormatFloat3;
        mtlVertexDescriptor.attributes[ModelVertexAttributePosition].offset = 0;
        mtlVertexDescriptor.attributes[ModelVertexAttributePosition].bufferIndex = ModelVertexInputIndexPosition;
        
        // texture coordinate
        mtlVertexDescriptor.attributes[ModelVertexAttributeTexcoord].format = MTLVertexFormatFloat2;
        mtlVertexDescriptor.attributes[ModelVertexAttributeTexcoord].offset = 16;
        mtlVertexDescriptor.attributes[ModelVertexAttributeTexcoord].bufferIndex = ModelVertexInputIndexPosition;
        
        // normal
        mtlVertexDescriptor.attributes[ModelVertexAttributeNormal].format = MTLVertexFormatFloat3;
        mtlVertexDescriptor.attributes[ModelVertexAttributeNormal].offset = 32;
        mtlVertexDescriptor.attributes[ModelVertexAttributeNormal].bufferIndex = ModelVertexInputIndexPosition;
        
        // layout
        mtlVertexDescriptor.layouts[ModelVertexInputIndexPosition].stride = 48;
        mtlVertexDescriptor.layouts[ModelVertexInputIndexPosition].stepRate = 1;
        mtlVertexDescriptor.layouts[ModelVertexInputIndexPosition].stepFunction = MTLVertexStepFunctionPerVertex;
        
        NSDictionary *attributeMap = @{
            [NSNumber numberWithInt:ModelVertexAttributePosition]: MDLVertexAttributePosition,
            [NSNumber numberWithInt:ModelVertexAttributeTexcoord]: MDLVertexAttributeTextureCoordinate,
            [NSNumber numberWithInt:ModelVertexAttributeNormal]: MDLVertexAttributeNormal
        };
        
        NSError *error;
        
        NSURL *foxJsonUrl = [[NSBundle common] URLForResource:@"fox.json"
                                                withExtension:nil
                                                 subdirectory:@"fox"];
        NSURL *foxTextureUrl = [[NSBundle common] URLForResource:@"fox.jpg"
                                                   withExtension:nil
                                                    subdirectory:@"fox"];
        _foxMesh = [[JsonMesh alloc] initWithJson:foxJsonUrl
                                      withTexture:foxTextureUrl
                                           device:_device
                              mtlVertexDescriptor:mtlVertexDescriptor
                                     attributesMap:attributeMap
                                            error:&error];
        NSAssert(_foxMesh, @"fox mesh error: %@", error);
        
        id<MTLLibrary> library = [device newDefaultLibrary];
        id<MTLFunction> vertexFunc = [library newFunctionWithName:@"vertexShader"];
        id<MTLFunction> fragmentFunc = [library newFunctionWithName:@"fragmentShader"];
        MTLRenderPipelineDescriptor *pipelineStateDescriptor = [[MTLRenderPipelineDescriptor alloc] init];
        pipelineStateDescriptor.label = @"Model pipeline";
        pipelineStateDescriptor.vertexDescriptor = mtlVertexDescriptor;
        pipelineStateDescriptor.vertexFunction = vertexFunc;
        pipelineStateDescriptor.fragmentFunction = fragmentFunc;
        pipelineStateDescriptor.sampleCount = mtkView.sampleCount;
        pipelineStateDescriptor.colorAttachments[0].pixelFormat = mtkView.colorPixelFormat;
        pipelineStateDescriptor.depthAttachmentPixelFormat = mtkView.depthStencilPixelFormat;
        
        _pipelineState = [_device newRenderPipelineStateWithDescriptor:pipelineStateDescriptor error:&error];
        
        NSAssert(_pipelineState, @"Failed to create pipeline state: %@", error);
        
        _commandQueue = [_device newCommandQueue];
        
        _viewportSize = (vector_uint2){ (int) (mtkView.frame.size.width), (int) (mtkView.frame.size.height) };
        float width = mtkView.frame.size.width;
        float height = mtkView.frame.size.height;
        
        _uniforms.modelMatrix = matrix4x4_translation(0.0, -1.5, 0.0);
        _uniforms.viewMatrix = [_camera getViewMatrix];
        _uniforms.projectionMatrix = matrix_perspective_left_hand(M_PI / 4.0, width / height, 0.1, 100.0);
    }
    
    return self;
}

- (void) handleMouseScrollDeltaX:(float) deltaX deltaY:(float) deltaY {
    
    if (fabsf(deltaX) >= fabsf(deltaY)) {
        [_satelliteController rotateCameraAroundTargetWithDeltaPhi:deltaX deltaTheta:0.0];
    } else {
        [_satelliteController rotateCameraAroundTargetWithDeltaPhi:0.0 deltaTheta:deltaY * 0.2];
    }
    
    _uniforms.viewMatrix = [_camera getViewMatrix];
}

- (void)drawInMTKView:(nonnull MTKView *)view {
    id<MTLCommandBuffer> commandBuffer = [_commandQueue commandBuffer];
    commandBuffer.label = @"Model Command";
    
    MTLRenderPassDescriptor *descriptor = [view currentRenderPassDescriptor];
    
    id<MTLRenderCommandEncoder> renderEncoder = [commandBuffer renderCommandEncoderWithDescriptor:descriptor];
    [renderEncoder setLabel:@"Model RenderEncoder"];
        
    [renderEncoder setViewport:(MTLViewport){ 0.0, 0.0, _viewportSize.x, _viewportSize.y, 0.0, 1.0 }];
    [renderEncoder setRenderPipelineState:_pipelineState];
    [renderEncoder setDepthStencilState:_depthState];
    
    [_foxMesh setVertexMeshToRenderEncoder:renderEncoder index:ModelVertexInputIndexPosition];
    
    [renderEncoder setVertexBytes:&_uniforms length:sizeof(_uniforms) atIndex:ModelVertexInputIndexUniforms];
    
    [_foxMesh drawMeshToRenderEncoder:renderEncoder textureHandler:^(MDLMaterialSemantic type, id<MTLTexture> texture, NSString* _) {
        if (type == MDLMaterialSemanticBaseColor) {
            [renderEncoder setFragmentTexture:texture atIndex:FragmentInputIndexDiffuseTexture];
        }
    }];
    
    [renderEncoder endEncoding];
    [commandBuffer presentDrawable:view.currentDrawable];
    [commandBuffer commit];
    [commandBuffer waitUntilCompleted];
}

- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size {
    _viewportSize.x = size.width;
    _viewportSize.y = size.height;
 
    _uniforms.projectionMatrix = matrix_perspective_left_hand(M_PI / 4.0, size.width / size.height, 0.1, 100.0);
}

@end

File: Asteroids/Asteroids/Renderer.swift
//
//  Renderer.swift
//  Asteroids
//
//  Created by Jacob Su on 3/28/21.
//

import Foundation
import MetalKit
import common
import ModelIO

class Renderer: NSObject {
    
    private var device: MTLDevice!
    private var camera: Camera!
    private var depthStencilState: MTLDepthStencilState!
    private var renderPipelineState: MTLRenderPipelineState!
    private var rockPipelineState: MTLRenderPipelineState!
    private var planetMesh: MetalMesh!
    private var rockMesh: MetalMesh!
    private var rockModelsBuffer: MTLBuffer!
    private var commandQueue: MTLCommandQueue!
    private var viewPort: MTLViewport!
    private var uniforms: Uniforms!
    private var rockUniform: RockUniforms!
    private let rocksAmount: Int = 10000
    
    init(metalView: MTKView) {
        super.init()
        
        camera = CameraFactory.generateRoundOrbitCamera(withPosition: vector_float3(0.0, 0.0, 155.0),
                                                        target: vector_float3(0.0, 0.0, 0.0),
                                                        up: vector_float3(0.0, 1.0, 0.0))
        
        device = metalView.device!

        metalView.delegate = self
        metalView.depthStencilPixelFormat = .depth32Float
        metalView.clearDepth = 1.0
        
        let depthStencilDescriptor = MTLDepthStencilDescriptor()
        depthStencilDescriptor.depthCompareFunction = .less
        depthStencilDescriptor.isDepthWriteEnabled = true
        
        depthStencilState = device.makeDepthStencilState(descriptor: depthStencilDescriptor)!
        
        let mtlVertexDescriptor = MTLVertexDescriptor()
        // position
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributePosition.rawValue)].format = .float3
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributePosition.rawValue)].offset = 0
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributePosition.rawValue)].bufferIndex = Int(ModelVertexInputIndexPosition.rawValue);
        
        // texture coordinate
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributeTexcoord.rawValue)].format = .float2
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributeTexcoord.rawValue)].offset = 16
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributeTexcoord.rawValue)].bufferIndex = Int(ModelVertexInputIndexPosition.rawValue)
        
        // normal
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributeNormal.rawValue)].format = .float3
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributeNormal.rawValue)].offset = 32
        mtlVertexDescriptor.attributes[Int(ModelVertexAttributeNormal.rawValue)].bufferIndex = Int(ModelVertexInputIndexPosition.rawValue)
        
        // layout
        mtlVertexDescriptor.layouts[Int(ModelVertexInputIndexPosition.rawValue)].stride = 48
        mtlVertexDescriptor.layouts[Int(ModelVertexInputIndexPosition.rawValue)].stepRate = 1
        mtlVertexDescriptor.layouts[Int(ModelVertexInputIndexPosition.rawValue)].stepFunction = .perVertex

        let planetUrl = Bundle.common.url(forResource: "planet.obj", withExtension: nil, subdirectory: "planet")!

        planetMesh = try! MetalMesh(withUrl: planetUrl,
                               device: device,
                               mtlVertexDescriptor: mtlVertexDescriptor,
                               attributeMap: [
                                Int(ModelVertexAttributePosition.rawValue) : MDLVertexAttributePosition,
                                Int(ModelVertexAttributeTexcoord.rawValue) : MDLVertexAttributeTextureCoordinate,
                                Int(ModelVertexAttributeNormal.rawValue)   : MDLVertexAttributeNormal
                               ])
        
        let rockUrl = Bundle.common.url(forResource: "rock.obj", withExtension: nil, subdirectory: "rock")!
        rockMesh = try! MetalMesh(withUrl: rockUrl,
                                  device: device,
                                  mtlVertexDescriptor: mtlVertexDescriptor,
                                  attributeMap: [
                                    Int(ModelVertexAttributePosition.rawValue) : MDLVertexAttributePosition,
                                    Int(ModelVertexAttributeTexcoord.rawValue) : MDLVertexAttributeTextureCoordinate,
                                    Int(ModelVertexAttributeNormal.rawValue)   : MDLVertexAttributeNormal])
        
        let library = device.makeDefaultLibrary()!
        let vertexFunc = library.makeFunction(name: "vertexShader")!
        let fragmentFunc = library.makeFunction(name: "fragmentShader")!
        let rockVertexFunc = library.makeFunction(name: "rockVertexShader")!
        
        let renderPipelineDescriptor = MTLRenderPipelineDescriptor()
        renderPipelineDescriptor.label = "model"
        renderPipelineDescriptor.vertexDescriptor = mtlVertexDescriptor
        renderPipelineDescriptor.vertexFunction = vertexFunc
        renderPipelineDescriptor.fragmentFunction = fragmentFunc
        renderPipelineDescriptor.colorAttachments[0].pixelFormat = metalView.colorPixelFormat
        renderPipelineDescriptor.depthAttachmentPixelFormat = metalView.depthStencilPixelFormat
        
        renderPipelineState = try! device.makeRenderPipelineState(descriptor: renderPipelineDescriptor)
        
        let rockRenderPipelineDescriptor = MTLRenderPipelineDescriptor()
        rockRenderPipelineDescriptor.label = "model"
        rockRenderPipelineDescriptor.vertexDescriptor = mtlVertexDescriptor
        rockRenderPipelineDescriptor.vertexFunction = rockVertexFunc
        rockRenderPipelineDescriptor.fragmentFunction = fragmentFunc
        rockRenderPipelineDescriptor.colorAttachments[0].pixelFormat = metalView.colorPixelFormat
        rockRenderPipelineDescriptor.depthAttachmentPixelFormat = metalView.depthStencilPixelFormat
        
        rockPipelineState = try! device.makeRenderPipelineState(descriptor: rockRenderPipelineDescriptor)
    
        commandQueue = device.makeCommandQueue()!
        
        viewPort = MTLViewport(originX: 0.0, originY: 0.0,
                               width: Double(metalView.frame.width),
                               height: Double(metalView.frame.height),
                               znear: 0.0, zfar: 1.0)
        
        let width : Float = Float(metalView.frame.width)
        let height: Float = Float(metalView.frame.height)
        
        uniforms = Uniforms(modelMatrix: matrix_multiply(matrix4x4_translation(0.0, -3.0, 0.0),
                                                         matrix4x4_scale(4.0, 4.0, 4.0)),
                            viewMatrix: camera.getViewMatrix(),
                            projectionMatrix: matrix_perspective_left_hand(Float.pi / 4.0, width / height, 0.1, 1000.0))
        rockUniform = RockUniforms(viewMatrix: camera.getViewMatrix(),
                                   projectionMatrix: matrix_perspective_left_hand(Float.pi / 4.0, width / height, 0.1, 1000.0))
        fillModelsBuffer()
    }
    
    func handleCameraEvent(deltaX: Float, deltaY: Float) -> Void {
        camera.rotateCameraAroundTarget(withDeltaPhi: deltaX * 0.2, deltaTheta: deltaY * 0.2)
        
        uniforms.viewMatrix = camera.getViewMatrix()
        rockUniform.viewMatrix = camera.getViewMatrix()
    }
    
    private func fillModelsBuffer() {
        var models : [RockModel] = [RockModel]()
        let radius : Float = 150.0
        let offset : Float = 25.0
        for i in 0..<rocksAmount {
            let angle = Float(i) / Float(rocksAmount) * 360.0
            var displacement = Float(Int.random(in: 0..<(2 * Int(offset) * 100))) / 100.0 - offset
            let x = sin(angle) * radius + displacement
            displacement = Float(Int.random(in: 0..<(2 * Int(offset) * 100))) / 100.0 - offset
            let y = displacement * 0.4
            displacement = Float(Int.random(in: 0..<(2 * Int(offset) * 100))) / 100.0 - offset
            let z = cos(angle) * radius + displacement
            
            let scale = Float(Int.random(in: 0..<40)) / 100.0 + 0.05
            let rotAngle = Float(Int.random(in: 0..<360)) / 360.0 * 2.0 * Float.pi
            
            let model = matrix_multiply(matrix4x4_translation(x, y, z),
                                        matrix_multiply(matrix4x4_scale(scale, scale, scale),
                                                        matrix4x4_rotation(rotAngle, vector_float3(0.4, 0.6, 0.8))))
            models.append(RockModel(modelMatrix: model))
        }
        
        let tmpBuffer = device.makeBuffer(bytes: models,
                                             length: MemoryLayout<RockModel>.stride * models.count,
                                             options: .storageModeShared)!
        rockModelsBuffer = device.makeBuffer(length: tmpBuffer.length, options: .storageModePrivate)!
        
        rockModelsBuffer.label = "Rocks Models"
        
        let commandBuffer = commandQueue.makeCommandBuffer()!
        let blitEncoder = commandBuffer.makeBlitCommandEncoder()!
        blitEncoder.copy(from: tmpBuffer, sourceOffset: 0, to: rockModelsBuffer, destinationOffset: 0, size: tmpBuffer.length)
        blitEncoder.endEncoding()
        commandBuffer.commit()
        commandBuffer.waitUntilCompleted()
    }
}

extension Renderer : MTKViewDelegate {
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        viewPort.width = Double(size.width)
        viewPort.height = Double(size.height)
        
        uniforms.projectionMatrix =
            matrix_perspective_left_hand(Float.pi / 4.0,
                                         Float(size.width) / Float(size.height),
                                         0.1,
                                         1000.0)
        rockUniform.projectionMatrix = uniforms.projectionMatrix
    }
    
    func draw(in view: MTKView) {
        let commandBuffer = commandQueue.makeCommandBuffer()!
        
        let renderDescriptor = view.currentRenderPassDescriptor!
        let renderEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: renderDescriptor)!
        
        renderEncoder.setViewport(viewPort)
        // draw planet
        renderEncoder.setRenderPipelineState(renderPipelineState)
        renderEncoder.setDepthStencilState(depthStencilState)
        
        renderEncoder.setVertexMesh(planetMesh, index: Int(ModelVertexInputIndexPosition.rawValue))
        
        renderEncoder.setVertexBytes(&uniforms,
                                     length: MemoryLayout<Uniforms>.stride,
                                     index: Int(ModelVertexInputIndexUniforms.rawValue))
        renderEncoder.drawMesh(planetMesh, textureHandler: { (type, texture, _) -> Void in
            if type == .baseColor {
                renderEncoder.setFragmentTexture(texture, index: Int(FragmentInputIndexDiffuseTexture.rawValue))
            }
        })
        
        // draw rocks
        renderEncoder.setRenderPipelineState(rockPipelineState)

        renderEncoder.setVertexMesh(rockMesh, index: Int(ModelVertexInputIndexPosition.rawValue))
        renderEncoder.setVertexBytes(&rockUniform, length: MemoryLayout<RockUniforms>.stride, index: Int(ModelVertexInputIndexUniforms.rawValue))
        renderEncoder.setVertexBuffer(rockModelsBuffer, offset: 0, index: Int(ModelVertexInputIndexModels.rawValue))

        renderEncoder.drawMesh(rockMesh, instanceCount: rocksAmount, textureHandler: { (type, texture, _) ->
            Void in
            if type == .baseColor {
                renderEncoder.setFragmentTexture(texture, index: Int(FragmentInputIndexDiffuseTexture.rawValue))
            }
        })

        renderEncoder.endEncoding()
        
        commandBuffer.present(view.currentDrawable!)
        commandBuffer.commit()
        commandBuffer.waitUntilCompleted()
    }
}
